ECOM 2764 course data input fields

AlasdairSwan opened the pull request:
@rlucioni @jimabramson please review. Not that once the APIs and all of the other  sub-tasks have been completed the tests will be more fully updated so increase coverage.

jimabramson commented on programs/static/js/models/course_model.js:
```diff
@@ -0,0 +1,79 @@
+                }
+
+                return config;
+            },
```
this looks familiar.  should it be moved into a base class?

AlasdairSwan replied:
Possibly, this was a copy & paste job and the other subtasks are going to tie it up to an API so it's kind of a WIP PR. I suggest waiting until the full implementation is finished. I could just delete the unused parts of this model for now if that would be cleaner?

jimabramson replied:
not a major issue for me if this is still WIP, we can clean it up when dependencies land & are integrated.   by the same token if it's not relevant to the changes you want reviewed, please feel free to remove it :-)

AlasdairSwan commented:
@jimabramson I cleaned up the model as discussed.

rlucioni commented on programs/static/templates/course_details.underscore:
```diff
@@ -22,11 +22,24 @@
+    <form class="form js-course-form">
+        <fieldset class="form-group">
+            <div class="field">
+                <label class="field-label" for="course-key-<%- cid %>"><%= gettext('Course Key') %></label>
```
Relatively minor, but this should be a "Course Code" instead of a "Course Key." I'd change the class name to reflect that, too.

rlucioni commented on programs/static/templates/course_details.underscore:
```diff
@@ -22,11 +22,24 @@
+                </div>
+            </div>
+            <div class="field">
+                <label class="field-label" for="course-key-<%- cid %>"><%= gettext('Course Title') %></label>
```
The `for` parameter should be `course-title`, right?

rlucioni commented on programs/static/templates/course_details.underscore:
```diff
@@ -22,11 +22,24 @@
+                    <span class="field-message-content"></span>
+                </div>
+            </div>
+            <button class="btn btn-primary js-select-course"><%- gettext('Select Course') %></button>
```
s/Select/Add?

AlasdairSwan replied:
There is already a button on the page saying 'Add a course'. Probably "Save Course' is better. @marcotuts thoughts?

jimabramson replied:
my $.02 is on 'save'

rlucioni replied:
Save works for me.

rlucioni commented on programs/static/js/views/course_details_view.js:
```diff
@@ -28,30 +31,36 @@ define([
                 this.$parentEl = $( this.parentEl );
                 this.programStatus = options.programStatus;
 
                 // Data passed in because it is not an actual model
```
Is this still true?

AlasdairSwan replied:
Yes it is

jimabramson commented:
@AlasdairSwan @marcotuts while we're allowing this input as free text i think some help text would be good to explain what these values mean/do.

for the key part, it might be best to stay consistent with Studio's messaging for the same element: "The unique number that identifies your course within your organization, e.g. CS101"  This *should* match the value in course runs in the LMS (though in practice our course run ids are not that clean, so it's not going to be enforced).

for the display name part the import thing is that this value is going to override whatever title might be set for the individual runs of the course, and will be displayed on the xseries progress page and in marketing presentations.  If the course runs' individual titles are changed it won't automatically be reflected here.

AlasdairSwan commented:
@jimabramson @rlucioni I have updated my PR in light of the review comments and added text for the field instructions. @marcotuts please review (and then offer improved versions of) the form field instructional text.

rlucioni commented on programs/static/js/models/course_model.js:
```diff
@@ -0,0 +1,25 @@
+            validation: {
+                course_code: {
+                    required: true,
+                    maxLength: 255
```
The database restricts this to 64 characters.

rlucioni commented on programs/static/js/models/course_model.js:
```diff
@@ -0,0 +1,25 @@
+                },
+                course_title: {
+                    required: true,
+                    maxLength: 255
```
The database restricts this to 128 characters.

jimabramson commented on programs/static/js/models/course_model.js:
```diff
@@ -0,0 +1,25 @@
+                    required: true,
+                    maxLength: 128
+                }
+            }
```
On the server this model is called CourseCode and the fields are 'key' and 'display_name'.  Would recommend staying closer to the same nomenclature if possible.

To work with the recent API changes this model is going to need another field 'organization' which I'm pretty sure needs to be a nested model.  The nested model won't/shouldn't change so you don't have to worry about updates to it, but it needs to be sent along with the course code object for creation/update/deletion.

jimabramson replied:
When creating the organization is always going to match the organization associated with the program so if necessary you can copy its values from there.

rlucioni commented on programs/static/js/views/course_details_view.js:
```diff
@@ -28,30 +31,36 @@ define([
                 }
 
-                this.model.on('change:display_name', this.update, this);
+                //this.programModel.on('change:display_name', this.update, this);
```
Should this be commented out?

AlasdairSwan replied:
It's to get test coverage to not be jacked - my next PR after this implements it properly.

rlucioni commented:
:+1:

jimabramson commented on programs/static/js/views/course_details_view.js:
```diff
@@ -28,30 +31,36 @@ define([
 
             postRender: function() {
-                var runs = this.model.get('run_modes');
+                var runs = this.programModel.get('run_modes');
```
i'm not sure if i should be ignoring this (for now) but in this and other places, it's implies that run_modes hang directly off the program in the data model, which isn't the case.  a program has many course codes, and each course code has many run modes.

jimabramson commented:
:+1: after rebasing.   Per discussion with @AlasdairSwan offline, we will deal with data model impedance mismatches in the next task/PR.

AlasdairSwan commented:
To make progress I am going to just merge this copy because the app is not on production yet and we can make any changes after Sylvia's review in a later PR.
